### DISCUSSION TITLE
XSLT error

### DISCUSSION THREAD

**Discussion by Parag411** [nextgenhealthcare/connect#5459] [General]
Hello Guys,

I am trying to use XSL template in Mirth XSLT transformer.
But I receive below error while validating the transformer:
<img width="556" alt="image" src="https://user-images.githubusercontent.com/113354874/194573134-86ff3aea-1064-45a8-8488-fae2b80e6ff9.png">

Below is the code in XSL for which I am receiving error:
    </xsl:template>
   <style type="text/css">
         <xsl:text>
body {
  color: #1f4a3e;
  background-color: #ffffff;
  font-family: Verdana, Tahoma, sans-serif;
  font-size: 11px;
}

a {
  color: #1f4a3e;
  background-color: #ffffff;
}

h1 {
  font-size: 12pt;
  font-weight: bold;
}

h2 {
  font-size: 11pt;
  font-weight: bold;
}

h3 {
  font-size: 10pt;
  font-weight: bold;
}

h4 {
  font-size: 8pt;
  font-weight: bold;
}

table {
  line-height: 10pt;
  width: 100%;
}

th {
  background-color: #1eb79a;
  color: #ffffff;
}

td {
  padding: 0.1cm 0.2cm;
  vertical-align: top;
  background-color: #cec9c9;
  color: #333333;
}

.h1center {
  font-size: 12pt;
  font-weight: bold;
  text-align: center;
  width: 80%;
}

.header_table{
  border: 0.5pt solid #00bfa5;
}

.td_label{
  font-weight: bold;
  color: white;
}

.td_header_role_name{
  width: 20%;
  background-color: #1eb79a;
  color: #ffffff;
}

.td_header_role_value{
  width: 80%;
  background-color: #cec9c9;
  color: #343433;
}

.Bold{
  font-weight: bold;
}

.Italics{
  font-style: italic;
}

.Underline{
  text-decoration:underline;
}

          </xsl:text>
        </style>
    </xsl:template>
</xsl:stylesheet>

Could you please let me know what is wrong in XSL file ?
I tried to validate the XSL file on online validators but they wont return specific error.
Thank you

**Comment by siddharthsharma9258** [comment]
That's CSS text which is throwing error. You don't need stylesheets in your xslt. Remove that block and try again.

**Reply by Parag411** [reply]
I tried it, the PDF output is distorted (not in proper format)

**Comment by jonbartels** [comment]
Please post the XSLT as a file attachment so the formatting doesn't get mangled.
If practical - please post the XSLT and a minimal input file that shows the error

**Reply by Parag411** [reply]
Sure, PFA XSL file (replace extension from .txt to .xsl)
I am unable to validate it in Mirth so unable to process the input file
[sample-ccda.txt](https://github.com/nextgenhealthcare/connect/files/9756189/sample-ccda.txt)

Thanks

**Reply by ab-20-m** [reply]
I believe the issue is caused because of how E4X interprets braces in XML initalisers.

From [ECMA-357 2nd Edition Section 8.3 XML Initialiser Input Elements](https://www.ecma-international.org/wp-content/uploads/ECMA-357_2nd_edition_december_2005.pdf)
> The left curly brace (**{**) and right curly brace (**}**) are used to delimit expressions that may be embedded in tags or 
element content to dynamically compute portions of the XML initialiser. The curly braces may appear in literal form 
inside an attribute value, a CDATA, PI, or XML Comment. In all other cases, the character reference **`&#x7B;`** shall 
be used to represent the left curly brace (**{**) and the character reference **`&#x7D;`** shall be used to represent the 
right curly brace (**}**).

**Reply by pacmano1** [reply]
In fact what @ab-20-m metioned is very handy for constructing things where you intenttionally want the insert variable names in to the XML.   For example

```JS
var stuff = 'localvarstuff'
$c('other_stuff','channelMap_var_stuff')

var sql = <>
select * from sometable where a_column = '{stuff}' and b_column = '{$c('other_stuff')}'
<>.toString();
logger.info(sql) 
```
would log to the console:
```
select * from sometable where a_column = 'localvarstuff' and b_column = 'channelMap_var_stuff'

```

Thanks to @tonygermano for sharing this technique a long time ago

**Reply by jonbartels** [reply]
The stylesheet posted is derived from the old Tyylitiedosto and Calvin Beebe stylesheet.

Some searches that might show working code for the same XLST might be:
* https://forums.mirthproject.io/search?q=cda+stylesheet&searchJSON=%7B%22keywords%22%3A%22cda+stylesheet%22%7D
* https://github.com/nextgenhealthcare/connect/search?q=cda+stylesheet&type=discussions

**Reply by pacmano1** [reply]
If you are using a document writer to attempt to take the transformed CDA to write to file, you will face many problems I suspect, like the fact that tranformed document which will render in a browser conflicts with the document writer expectation that all tags are closed like `<br>`.  You will also need to remove META tags, DOCTYPE tags and so on.

**Reply by Parag411** [reply]
> &#x7D;

Thanks @ab-20-m 
It resolved the error related to curly braces.

Thanks alot.

**Reply by Parag411** [reply]
> If you are using a document writer to attempt to take the transformed CDA to write to file, you will face many problems I suspect, like the fact that tranformed document which will render in a browser conflicts with the document writer expectation that all tags are closed like `<br>`. You will also need to remove META tags, DOCTYPE tags and so on.

You are right @pacmano1 , I am yet to reach to that step.
